File: server/src/main/kotlin/com/smartclock/server/service/SyncService.kt
package com.smartclock.server.service

import com.smartclock.server.dto.AlarmLogAdminDto
import com.smartclock.server.dto.AlarmLogBatchRequest
import com.smartclock.server.dto.AlarmLogBatchResponse
import com.smartclock.server.dto.AlarmPullResponse
import com.smartclock.server.dto.AlarmPushRequest
import com.smartclock.server.dto.AlarmPushResponse
import com.smartclock.server.dto.AlarmSyncDto
import com.smartclock.server.dto.ApiUserDto
import com.smartclock.server.dto.BootstrapResponse
import com.smartclock.server.dto.toDto
import com.smartclock.server.model.UserProfile
import com.smartclock.server.support.ApiException
import com.smartclock.server.support.toInstantUtc
import com.smartclock.server.support.toIsoUtc
import com.smartclock.server.support.toSqlTimestamp
import java.sql.Timestamp
import java.time.Clock
import java.time.Instant
import org.jooq.DSLContext
import org.jooq.Record
import org.jooq.impl.DSL
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service

@Service
class SyncService(
    private val dsl: DSLContext,
    private val authService: AuthService,
    private val clock: Clock
) {

    fun bootstrap(userId: Long): BootstrapResponse {
        val user = authService.me(com.smartclock.server.security.AuthUserPrincipal(userId, com.smartclock.server.model.Role.USER))
        val alarms = dsl.fetch(
            "SELECT * FROM alarms WHERE user_id = ? ORDER BY updated_at ASC",
            userId
        ).map(::mapAlarm)
        return BootstrapResponse(
            user = user.toDto(),
            alarms = alarms,
            serverTime = Instant.now(clock).toIsoUtc()
        )
    }

    fun pushAlarms(userId: Long, request: AlarmPushRequest): AlarmPushResponse = dsl.transactionResult { cfg ->
        val tx = DSL.using(cfg)
        val results = request.alarms.map { upsertAlarm(tx, userId, it) }
        AlarmPushResponse(results, Instant.now(clock).toIsoUtc())
    }

    fun pullAlarms(userId: Long, since: String?): AlarmPullResponse {
        val alarms = if (since.isNullOrBlank()) {
            dsl.fetch(
                "SELECT * FROM alarms WHERE user_id = ? ORDER BY updated_at ASC",
                userId
            )
        } else {
            dsl.fetch(
                "SELECT * FROM alarms WHERE user_id = ? AND updated_at > ? ORDER BY updated_at ASC",
                userId,
                since.toInstantUtc().toSqlTimestamp()
            )
        }.map(::mapAlarm)
        return AlarmPullResponse(alarms = alarms, serverTime = Instant.now(clock).toIsoUtc())
    }

    fun uploadLogs(userId: Long, request: AlarmLogBatchRequest): AlarmLogBatchResponse = dsl.transactionResult { cfg ->
        val tx = DSL.using(cfg)
        var inserted = 0
        var ignored = 0
        val now = Instant.now(clock).toSqlTimestamp()
        request.alarmLogs.forEach { log ->
            val rows = tx.execute(
                """
                INSERT IGNORE INTO alarm_logs
                (alarm_id, user_id, fired_at, action, device_id, log_hash, created_at)
                VALUES (?, ?, ?, ?, ?, ?, ?)
                """.trimIndent(),
                log.alarmId,
                userId,
                log.firedAt.toInstantUtc().toSqlTimestamp(),
                log.action,
                log.deviceId,
                log.logHash,
                now
            )
            if (rows > 0) inserted++ else ignored++
        }
        AlarmLogBatchResponse(inserted = inserted, ignored = ignored, serverTime = Instant.now(clock).toIsoUtc())
    }

    private fun upsertAlarm(tx: DSLContext, userId: Long, dto: AlarmSyncDto): AlarmSyncDto {
        val now = Instant.now(clock)
        val existing = tx.fetchOne(
            "SELECT id, created_at FROM alarms WHERE user_id = ? AND client_uuid = ? LIMIT 1",
            userId,
            dto.clientUuid
        )

        val createdAt = dto.createdAt?.toInstantUtc() ?: now
        if (existing == null) {
            tx.execute(
                """
                INSERT INTO alarms
                (user_id, client_uuid, type, title, note, trigger_time, duration_sec, repeat_weekdays,
                 repeat_month_days, anniversary_calendar, advance_notify_days, ringtone, vibrate,
                 volume_fade, snooze_minutes, label, color, enabled, start_date, end_date,
                 schedule_mode, alert_policy, time_anchor_mode, interval_months, interval_years,
                 template_id, status, created_at, updated_at)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """.trimIndent(),
                userId,
                dto.clientUuid,
                dto.type,
                dto.title,
                dto.note,
                dto.triggerTime?.toInstantUtc()?.toSqlTimestamp(),
                dto.durationSec,
                dto.repeatWeekdays,
                dto.repeatMonthDays,
                dto.anniversaryCalendar,
                dto.advanceNotifyDays,
                dto.ringtone,
                if (dto.vibrate) 1 else 0,
                if (dto.volumeFade) 1 else 0,
                dto.snoozeMinutes,
                dto.label,
                dto.color,
                if (dto.enabled) 1 else 0,
                dto.startDate?.toInstantUtc()?.toSqlTimestamp(),
                dto.endDate?.toInstantUtc()?.toSqlTimestamp(),
                dto.scheduleMode,
                dto.alertPolicy,
                dto.timeAnchorMode,
                dto.intervalMonths.coerceAtLeast(1),
                dto.intervalYears.coerceAtLeast(1),
                dto.templateId,
                dto.status,
                createdAt.toSqlTimestamp(),
                now.toSqlTimestamp()
            )
            val serverId = (tx.fetchValue("SELECT LAST_INSERT_ID()") as? Number)?.toLong()
                ?: throw ApiException(HttpStatus.INTERNAL_SERVER_ERROR, "failed to insert alarm")
            return fetchAlarmById(tx, serverId)
        }

        val alarmId = existing.get("id", Long::class.java)!!
        tx.execute(
            """
            UPDATE alarms SET
              type = ?, title = ?, note = ?, trigger_time = ?, duration_sec = ?, repeat_weekdays = ?,
              repeat_month_days = ?, anniversary_calendar = ?, advance_notify_days = ?, ringtone = ?,
              vibrate = ?, volume_fade = ?, snooze_minutes = ?, label = ?, color = ?, enabled = ?,
              start_date = ?, end_date = ?, schedule_mode = ?, alert_policy = ?, time_anchor_mode = ?,
              interval_months = ?, interval_years = ?, template_id = ?, status = ?, updated_at = ?
            WHERE id = ? AND user_id = ?
            """.trimIndent(),
            dto.type,
            dto.title,
            dto.note,
            dto.triggerTime?.toInstantUtc()?.toSqlTimestamp(),
            dto.durationSec,
            dto.repeatWeekdays,
            dto.repeatMonthDays,
            dto.anniversaryCalendar,
            dto.advanceNotifyDays,
            dto.ringtone,
            if (dto.vibrate) 1 else 0,
            if (dto.volumeFade) 1 else 0,
            dto.snoozeMinutes,
            dto.label,
            dto.color,
            if (dto.enabled) 1 else 0,
            dto.startDate?.toInstantUtc()?.toSqlTimestamp(),
            dto.endDate?.toInstantUtc()?.toSqlTimestamp(),
            dto.scheduleMode,
            dto.alertPolicy,
            dto.timeAnchorMode,
            dto.intervalMonths.coerceAtLeast(1),
            dto.intervalYears.coerceAtLeast(1),
            dto.templateId,
            dto.status,
            now.toSqlTimestamp(),
            alarmId,
            userId
        )
        return fetchAlarmById(tx, alarmId)
    }

    fun fetchAlarmById(tx: DSLContext, alarmId: Long): AlarmSyncDto =
        tx.fetchOne("SELECT * FROM alarms WHERE id = ? LIMIT 1", alarmId)
            ?.let(::mapAlarm)
            ?: throw ApiException(HttpStatus.NOT_FOUND, "alarm not found")

    fun mapAlarm(record: Record): AlarmSyncDto = AlarmSyncDto(
        id = record.get("id", Long::class.java),
        clientUuid = record.get("client_uuid", String::class.java)!!,
        type = record.get("type", Int::class.java) ?: 0,
        title = record.get("title", String::class.java) ?: "",
        note = record.get("note", String::class.java),
        triggerTime = record.get("trigger_time", Timestamp::class.java)?.toInstant()?.toIsoUtc(),
        durationSec = record.get("duration_sec", Int::class.java),
        repeatWeekdays = record.get("repeat_weekdays", Int::class.java) ?: 0,
        repeatMonthDays = record.get("repeat_month_days", Int::class.java) ?: 0,
        anniversaryCalendar = record.get("anniversary_calendar", Int::class.java) ?: 0,
        advanceNotifyDays = record.get("advance_notify_days", String::class.java),
        ringtone = record.get("ringtone", String::class.java),
        vibrate = (record.get("vibrate", Int::class.java) ?: 0) == 1,
        volumeFade = (record.get("volume_fade", Int::class.java) ?: 0) == 1,
        snoozeMinutes = record.get("snooze_minutes", Int::class.java) ?: 5,
        label = record.get("label", String::class.java),
        color = record.get("color", String::class.java),
        enabled = (record.get("enabled", Int::class.java) ?: 0) == 1,
        startDate = record.get("start_date", Timestamp::class.java)?.toInstant()?.toIsoUtc(),
        endDate = record.get("end_date", Timestamp::class.java)?.toInstant()?.toIsoUtc(),
        scheduleMode = record.get("schedule_mode", Int::class.java) ?: 0,
        alertPolicy = record.get("alert_policy", Int::class.java) ?: 0,
        timeAnchorMode = record.get("time_anchor_mode", Int::class.java) ?: 0,
        intervalMonths = record.get("interval_months", Int::class.java) ?: 1,
        intervalYears = record.get("interval_years", Int::class.java) ?: 1,
        templateId = record.get("template_id", String::class.java),
        status = record.get("status", Int::class.java) ?: 0,
        createdAt = record.get("created_at", Timestamp::class.java)?.toInstant()?.toIsoUtc(),
        updatedAt = record.get("updated_at", Timestamp::class.java)?.toInstant()?.toIsoUtc()
    )
}

File: server-php/src/Service/SyncService.php
<?php

declare(strict_types=1);

namespace SmartClock\Server\Service;

use PDO;
use SmartClock\Server\Support\ApiException;
use SmartClock\Server\Support\Time;

final class SyncService
{
    public function __construct(
        private readonly PDO $pdo
    ) {
    }

    public function bootstrap(array $user): array
    {
        $stmt = $this->pdo->prepare(
            'SELECT * FROM alarms WHERE user_id = :user_id ORDER BY updated_at ASC'
        );
        $stmt->execute([':user_id' => (int) $user['id']]);

        return [
            'user' => [
                'id' => (int) $user['id'],
                'phone' => $user['phone'],
                'email' => $user['email'],
                'nickname' => $user['nickname'],
                'role' => $user['role'],
                'status' => (int) $user['status'],
            ],
            'alarms' => array_map(fn (array $row) => $this->mapAlarm($row), $stmt->fetchAll()),
            'serverTime' => Time::toIso(Time::now()),
        ];
    }

    public function pushAlarms(int $userId, array $payload): array
    {
        $alarms = $payload['alarms'] ?? null;
        if (!is_array($alarms) || $alarms === []) {
            throw new ApiException(400, 'alarms is required', 'bad_request');
        }

        $result = [];
        $errors = [];

        foreach ($alarms as $alarm) {
            if (!is_array($alarm)) {
                $errors[] = [
                    'clientUuid' => null,
                    'message' => 'alarm payload is invalid',
                ];
                continue;
            }

            $clientUuid = trim((string) ($alarm['clientUuid'] ?? ''));
            try {
                $result[] = $this->upsertAlarm($userId, $alarm);
            } catch (\Throwable $e) {
                $message = trim($e->getMessage()) !== '' ? $e->getMessage() : 'push alarm failed';
                $errors[] = [
                    'clientUuid' => $clientUuid !== '' ? $clientUuid : null,
                    'message' => $message,
                ];
                error_log(sprintf(
                    '[SmartClock PHP] push alarm failed user=%d client=%s %s in %s:%d',
                    $userId,
                    $clientUuid !== '' ? $clientUuid : '-',
                    $message,
                    $e->getFile(),
                    $e->getLine()
                ));
            }
        }

        if ($result === []) {
            $firstError = $errors[0]['message'] ?? 'push alarms failed';
            throw new ApiException(500, $firstError, 'server_error');
        }

        return [
            'alarms' => $result,
            'errors' => $errors,
            'serverTime' => Time::toIso(Time::now()),
        ];
    }

    public function pullAlarms(int $userId, ?string $since): array
    {
        if ($since === null || $since === '') {
            $stmt = $this->pdo->prepare(
                'SELECT * FROM alarms WHERE user_id = :user_id ORDER BY updated_at ASC'
            );
            $stmt->execute([':user_id' => $userId]);
        } else {
            $stmt = $this->pdo->prepare(
                'SELECT * FROM alarms WHERE user_id = :user_id AND updated_at > :since ORDER BY updated_at ASC'
            );
            $stmt->execute([
                ':user_id' => $userId,
                ':since' => Time::toDb(Time::parseIso($since)),
            ]);
        }

        return [
            'alarms' => array_map(fn (array $row) => $this->mapAlarm($row), $stmt->fetchAll()),
            'serverTime' => Time::toIso(Time::now()),
        ];
    }

    public function uploadLogs(int $userId, array $payload): array
    {
        $alarmLogs = $payload['alarmLogs'] ?? null;
        if (!is_array($alarmLogs) || $alarmLogs === []) {
            throw new ApiException(400, 'alarmLogs is required', 'bad_request');
        }

        $stmt = $this->pdo->prepare(
            'INSERT IGNORE INTO alarm_logs
             (alarm_id, user_id, fired_at, action, device_id, log_hash, created_at)
             VALUES (:alarm_id, :user_id, :fired_at, :action, :device_id, :log_hash, :created_at)'
        );

        $inserted = 0;
        $ignored = 0;
        $errors = [];
        $now = Time::toDb(Time::now());

        foreach ($alarmLogs as $log) {
            if (!is_array($log)) {
                $ignored++;
                $errors[] = [
                    'alarmId' => null,
                    'message' => 'alarm log payload is invalid',
                ];
                continue;
            }
            $alarmId = (int) ($log['alarmId'] ?? 0);
            $firedAt = (string) ($log['firedAt'] ?? '');
            $deviceId = trim((string) ($log['deviceId'] ?? ''));
            $logHash = trim((string) ($log['logHash'] ?? ''));

            if ($alarmId <= 0 || $firedAt === '' || $deviceId === '' || $logHash === '') {
                $ignored++;
                $errors[] = [
                    'alarmId' => $alarmId > 0 ? $alarmId : null,
                    'message' => 'invalid alarm log payload',
                ];
                continue;
            }

            try {
                $stmt->execute([
                    ':alarm_id' => $alarmId,
                    ':user_id' => $userId,
                    ':fired_at' => Time::toDb(Time::parseIso($firedAt)),
                    ':action' => (int) ($log['action'] ?? 0),
                    ':device_id' => $deviceId,
                    ':log_hash' => $logHash,
                    ':created_at' => $now,
                ]);

                if ($stmt->rowCount() > 0) {
                    $inserted++;
                } else {
                    $ignored++;
                }
            } catch (\Throwable $e) {
                $ignored++;
                $message = trim($e->getMessage()) !== '' ? $e->getMessage() : 'insert alarm log failed';
                $errors[] = [
                    'alarmId' => $alarmId,
                    'message' => $message,
                ];
                error_log(sprintf(
                    '[SmartClock PHP] upload alarm log failed user=%d alarm=%d %s in %s:%d',
                    $userId,
                    $alarmId,
                    $message,
                    $e->getFile(),
                    $e->getLine()
                ));
            }
        }

        return [
            'inserted' => $inserted,
            'ignored' => $ignored,
            'errors' => $errors,
            'serverTime' => Time::toIso(Time::now()),
        ];
    }

    public function getAlarmById(int $alarmId): array
    {
        $stmt = $this->pdo->prepare('SELECT * FROM alarms WHERE id = :id LIMIT 1');
        $stmt->execute([':id' => $alarmId]);
        $record = $stmt->fetch();
        if (!$record) {
            throw new ApiException(404, 'alarm not found', 'not_found');
        }

        return $record;
    }

    public function mapAlarm(array $record): array
    {
        return [
            'id' => isset($record['id']) ? (int) $record['id'] : null,
            'clientUuid' => (string) $record['client_uuid'],
            'type' => (int) $record['type'],
            'title' => (string) $record['title'],
            'note' => $record['note'] !== null ? (string) $record['note'] : null,
            'triggerTime' => Time::toIso(Time::parseDb($record['trigger_time'] ?? null)),
            'durationSec' => isset($record['duration_sec']) ? (int) $record['duration_sec'] : null,
            'repeatWeekdays' => (int) ($record['repeat_weekdays'] ?? 0),
            'repeatMonthDays' => (int) ($record['repeat_month_days'] ?? 0),
            'anniversaryCalendar' => (int) ($record['anniversary_calendar'] ?? 0),
            'advanceNotifyDays' => $record['advance_notify_days'] !== null ? (string) $record['advance_notify_days'] : null,
            'ringtone' => $record['ringtone'] !== null ? (string) $record['ringtone'] : null,
            'vibrate' => (int) ($record['vibrate'] ?? 0) === 1,
            'volumeFade' => (int) ($record['volume_fade'] ?? 0) === 1,
            'snoozeMinutes' => (int) ($record['snooze_minutes'] ?? 5),
            'label' => $record['label'] !== null ? (string) $record['label'] : null,
            'color' => $record['color'] !== null ? (string) $record['color'] : null,
            'enabled' => (int) ($record['enabled'] ?? 0) === 1,
            'startDate' => Time::toIso(Time::parseDb($record['start_date'] ?? null)),
            'endDate' => Time::toIso(Time::parseDb($record['end_date'] ?? null)),
            'scheduleMode' => (int) ($record['schedule_mode'] ?? 0),
            'alertPolicy' => (int) ($record['alert_policy'] ?? 0),
            'timeAnchorMode' => (int) ($record['time_anchor_mode'] ?? 0),
            'intervalMonths' => (int) ($record['interval_months'] ?? 1),
            'intervalYears' => (int) ($record['interval_years'] ?? 1),
            'nextOverrideMode' => (int) ($record['next_override_mode'] ?? 0),
            'nextOverrideAnchorDate' => $record['next_override_anchor_date'] !== null ? (string) $record['next_override_anchor_date'] : null,
            'nextOverrideAnchorTriggerAt' => Time::toIso(Time::parseDb($record['next_override_anchor_trigger_at'] ?? null)),
            'nextOverrideTriggerAt' => Time::toIso(Time::parseDb($record['next_override_trigger_at'] ?? null)),
            'templateId' => $record['template_id'] !== null ? (string) $record['template_id'] : null,
            'status' => (int) ($record['status'] ?? 0),
            'createdAt' => Time::toIso(Time::parseDb($record['created_at'] ?? null)),
            'updatedAt' => Time::toIso(Time::parseDb($record['updated_at'] ?? null)),
        ];
    }

    private function upsertAlarm(int $userId, array $dto): array
    {
        $clientUuid = trim((string) ($dto['clientUuid'] ?? ''));
        $title = $this->limitText((string) ($dto['title'] ?? ''), 128, true);
        if ($clientUuid === '' || $title === '') {
            throw new ApiException(400, 'clientUuid and title are required', 'bad_request');
        }

        $existingStmt = $this->pdo->prepare(
            'SELECT id, created_at FROM alarms WHERE user_id = :user_id AND client_uuid = :client_uuid LIMIT 1'
        );
        $existingStmt->execute([
            ':user_id' => $userId,
            ':client_uuid' => $clientUuid,
        ]);
        $existing = $existingStmt->fetch();

        $now = Time::now();
        $createdAt = Time::parseIso(isset($dto['createdAt']) ? (string) $dto['createdAt'] : null) ?? $now;
        $fields = [
            ':user_id' => $userId,
            ':client_uuid' => $clientUuid,
            ':type' => (int) ($dto['type'] ?? 0),
            ':title' => $title,
            ':note' => isset($dto['note']) ? (string) $dto['note'] : null,
            ':trigger_time' => Time::toDb(Time::parseIso(isset($dto['triggerTime']) ? (string) $dto['triggerTime'] : null)),
            ':duration_sec' => isset($dto['durationSec']) ? (int) $dto['durationSec'] : null,
            ':repeat_weekdays' => (int) ($dto['repeatWeekdays'] ?? 0),
            ':repeat_month_days' => (int) ($dto['repeatMonthDays'] ?? 0),
            ':anniversary_calendar' => (int) ($dto['anniversaryCalendar'] ?? 0),
            ':advance_notify_days' => $this->limitText(
                isset($dto['advanceNotifyDays']) ? (string) $dto['advanceNotifyDays'] : null,
                128,
                true
            ),
            ':ringtone' => $this->limitText(
                isset($dto['ringtone']) ? (string) $dto['ringtone'] : null,
                255,
                true
            ),
            ':vibrate' => !empty($dto['vibrate']) ? 1 : 0,
            ':volume_fade' => !empty($dto['volumeFade']) ? 1 : 0,
            ':snooze_minutes' => (int) ($dto['snoozeMinutes'] ?? 5),
            ':label' => $this->limitText(
                isset($dto['label']) ? (string) $dto['label'] : null,
                128,
                true
            ),
            ':color' => $this->limitText(
                isset($dto['color']) ? (string) $dto['color'] : null,
                32,
                true
            ),
            ':enabled' => !empty($dto['enabled']) ? 1 : 0,
            ':start_date' => Time::toDb(Time::parseIso(isset($dto['startDate']) ? (string) $dto['startDate'] : null)),
            ':end_date' => Time::toDb(Time::parseIso(isset($dto['endDate']) ? (string) $dto['endDate'] : null)),
            ':schedule_mode' => (int) ($dto['scheduleMode'] ?? 0),
            ':alert_policy' => (int) ($dto['alertPolicy'] ?? 0),
            ':time_anchor_mode' => (int) ($dto['timeAnchorMode'] ?? 0),
            ':interval_months' => max(1, (int) ($dto['intervalMonths'] ?? 1)),
            ':interval_years' => max(1, (int) ($dto['intervalYears'] ?? 1)),
            ':next_override_mode' => (int) ($dto['nextOverrideMode'] ?? 0),
            ':next_override_anchor_date' => $this->limitText(
                isset($dto['nextOverrideAnchorDate']) ? (string) $dto['nextOverrideAnchorDate'] : null,
                32,
                true
            ),
            ':next_override_anchor_trigger_at' => Time::toDb(Time::parseIso(isset($dto['nextOverrideAnchorTriggerAt']) ? (string) $dto['nextOverrideAnchorTriggerAt'] : null)),
            ':next_override_trigger_at' => Time::toDb(Time::parseIso(isset($dto['nextOverrideTriggerAt']) ? (string) $dto['nextOverrideTriggerAt'] : null)),
            ':template_id' => $this->limitText(
                isset($dto['templateId']) ? (string) $dto['templateId'] : null,
                64,
                true
            ),
            ':status' => (int) ($dto['status'] ?? 0),
            ':created_at' => Time::toDb($createdAt),
            ':updated_at' => Time::toDb($now),
        ];

        if (!$existing) {
            $stmt = $this->pdo->prepare(
                'INSERT INTO alarms
                (user_id, client_uuid, type, title, note, trigger_time, duration_sec, repeat_weekdays,
                 repeat_month_days, anniversary_calendar, advance_notify_days, ringtone, vibrate,
                 volume_fade, snooze_minutes, label, color, enabled, start_date, end_date,
                 schedule_mode, alert_policy, time_anchor_mode, interval_months, interval_years,
                 next_override_mode, next_override_anchor_date, next_override_anchor_trigger_at, next_override_trigger_at,
                 template_id, status, created_at, updated_at)
                 VALUES
                (:user_id, :client_uuid, :type, :title, :note, :trigger_time, :duration_sec, :repeat_weekdays,
                 :repeat_month_days, :anniversary_calendar, :advance_notify_days, :ringtone, :vibrate,
                 :volume_fade, :snooze_minutes, :label, :color, :enabled, :start_date, :end_date,
                 :schedule_mode, :alert_policy, :time_anchor_mode, :interval_months, :interval_years,
                 :next_override_mode, :next_override_anchor_date, :next_override_anchor_trigger_at, :next_override_trigger_at,
                 :template_id, :status, :created_at, :updated_at)'
            );
            $stmt->execute($fields);
            return $this->mapAlarm($this->getAlarmById((int) $this->pdo->lastInsertId()));
        }

        $stmt = $this->pdo->prepare(
            'UPDATE alarms SET
              type = :type, title = :title, note = :note, trigger_time = :trigger_time, duration_sec = :duration_sec,
              repeat_weekdays = :repeat_weekdays, repeat_month_days = :repeat_month_days, anniversary_calendar = :anniversary_calendar,
              advance_notify_days = :advance_notify_days, ringtone = :ringtone, vibrate = :vibrate, volume_fade = :volume_fade,
              snooze_minutes = :snooze_minutes, label = :label, color = :color, enabled = :enabled,
              start_date = :start_date, end_date = :end_date, schedule_mode = :schedule_mode, alert_policy = :alert_policy,
              time_anchor_mode = :time_anchor_mode, interval_months = :interval_months, interval_years = :interval_years,
              next_override_mode = :next_override_mode, next_override_anchor_date = :next_override_anchor_date,
              next_override_anchor_trigger_at = :next_override_anchor_trigger_at, next_override_trigger_at = :next_override_trigger_at,
              template_id = :template_id, status = :status, updated_at = :updated_at
             WHERE id = :id AND user_id = :user_id'
        );
        $stmt->execute($fields + [':id' => (int) $existing['id']]);

        return $this->mapAlarm($this->getAlarmById((int) $existing['id']));
    }

    private function limitText(?string $value, int $maxLength, bool $trim = false): ?string
    {
        if ($value === null) {
            return null;
        }

        $normalized = $trim ? trim($value) : $value;
        if ($normalized === '') {
            return $normalized;
        }

        if (function_exists('mb_substr')) {
            return mb_substr($normalized, 0, $maxLength, 'UTF-8');
        }

        if (preg_match_all('/./us', $normalized, $matches) === 1) {
            return implode('', array_slice($matches[0], 0, $maxLength));
        }

        return substr($normalized, 0, $maxLength);
    }
}
